sigsci_site_manager/util.py

```diff
@@ -0,0 +1,106 @@
+    else:
+        ret = {k: data[k] for k in keys}
+    return ret
+
+
+def equal_conditions(a, b):
+    # Check the simple fields first
+    if a['type'] != b['type']:
+        return False
+
+    if a['type'] == 'single':
+        # Single conditions have simple values
+        for key in ['field', 'operator', 'value']:
+            if a[key] != b[key]:
+                return False
+        return True
+
+    if a['type'] in ('group', 'multival'):
+        # groupOperator is part of both group and multival
+        if a['groupOperator'] != b['groupOperator']:
+            return False
+
+        if a['type'] == 'multival':
+            # multival has additional fields
+            for key in ['field', 'operator']:
+                if a[key] != b[key]:
+                    return False
+
+        # Make sure there are the same number of conditions before doing
+        # anything more complicated
+        if len(a['conditions']) != len(b['conditions']):
+            return False
+
+        # Loop through all in list a
+        match_all = False
+        for conda in a['conditions']:
+            # Compare the element in a against all in list b
+            for condb in b['conditions']:
+                # Check each field for a match
+                match = True
+                for key in ['type', 'field', 'operator', 'value']:
+                    if conda[key] != condb[key]:
+                        # Non-matching field found
+                        match = False
+                if match:
+                    # Found a match so stop looping through b
+                    break
+            else:
+                # Nothing in b matched the item from a so stop looping
+                break
+        else:
+            # Never broke from the loop of a due to not matches so all must
+            # have matched
+            match_all = True
+
+        return match_all
+
+
+def equal_rules(in_a, in_b, signal_rule=False):
+    """
+    Compares the details of a request or signal rule, exluding some fields
+    that should not be compared.
+    """
+    if signal_rule:
+        # Signal rules have an implied action
+        keys = ['groupOperator', 'conditions', 'signal']
+        simple_fields = ['groupOperator', 'signal']
+    else:
+        # Request rules have an explicit action
+        keys = ['groupOperator', 'conditions', 'action', 'signal']
+        simple_fields = ['groupOperator', 'action', 'signal']
+
+    a = filter_data(in_a, keys)
+    b = filter_data(in_b, keys)
+
+    # Check the simple fields first
+    for key in simple_fields:
+        if a[key] != b[key]:
+            return False
+
+    # Make sure there are the same number of conditions before doing anything
+    # more complicated
+    if len(a['conditions']) != len(b['conditions']):
+        return False
+
+    match_all = False
+    for conda in a['conditions']:
+        for condb in b['conditions']:
+            if equal_conditions(conda, condb):
+                break
```

Comment:
This doesn't look right, if condition 'a' and condition 'b' both match, we should break and leave ```match_all=False```?

Also, this could also be solved by straight up string comparison like in the equal_conditions comment I left above.

Reply:
There is still the list ordering problem for string-based comparisons.

If the item in `a` matches any item in `b` we break out of the `b` loop. If we never break out of the `b` loop we hit the `else` for the `b` loop, which means `a` didn't match anything in `b`, so break out of the `a` loop. If we never break out of the `a` loop it means the `a` item always broke out of the `b` loop on a match and so we hit the `else` for the `a` loop and set `match_all` to `True`, otherwise `match_all` stays `False`.